File: 016-three-sum-closest/solution.py
#自己寫的暴力解
from typing import List
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        n = len(nums)
        nums.sort()
        min_diff = float('inf')
        diff = 0
        ans = 0
        for l in range(0, n - 2):
            k = l+1
            r = n-1
            t = target - nums[l]
            while k < r:
                Sum = nums[k] + nums[r]
                if Sum < t:
                    diff = abs(Sum - t)
                    k += 1
                elif Sum > t:
                    diff = abs(Sum - t)
                    r -= 1
                else:
                    return nums[l] + Sum
                if min_diff > diff:
                    min_diff = diff
                    ans = nums[l] + Sum
        return ans
if __name__ == "__main__":
    nums = [4,0,5,-5,3,3,0,-4,-5]
    print("Ans = ", Solution().threeSumClosest(nums, -2))

    """
    建議先看過第15題
    015-three-sum/Solution.java

    指針示意圖
    t = -2 -(-5) = 3
     l   k                     r
    -5, -5, -4, 0, 0, 3, 3, 4, 5

     l       k                 r
    -5, -5, -4, 0, 0, 3, 3, 4, 5

    for 迴圈把 l 固定住, 利用 k, r 找 t
    while 迴圈中要盡力找到 nums[k] + nums[r] = t 
    如果沒有找到的話, 也要記錄下最接近的組合並保存到 diff 中
    """

File: 016-three-sum-closest/Solution.java
//嘗試優化python版本失敗, 但寫出了另一個版本
import java.util.Arrays;

class Solution16 {
    public static int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int n = nums.length;
        int minDiff = Integer.MAX_VALUE;
        int diff = 0;
        int ans = 0;
        for (int l = 0; l < n - 2; l++) {
            int k = l + 1;
            int r = n - 1;
            while (k + 1 !=  r){
                if (nums[l] + nums[k] + nums[r] == target)
                    return target;
                while (k + 1 != r && nums[l] + nums[k] + nums[r] < target){
                    diff = Math.abs(nums[l] + nums[k] + nums[r] - target);
                    if (minDiff > diff){
                        minDiff = diff;
                        ans = nums[l] + nums[k] + nums[r];
                    }
                    k++;
                }
                while (k + 1 !=  r && nums[l] + nums[k] + nums[r] > target){
                    diff = Math.abs(nums[l] + nums[k] + nums[r] - target);
                    if (minDiff > diff){
                        minDiff = diff;
                        ans = nums[l] + nums[k] + nums[r];
                    }
                    r--;
                }
            }
            diff = Math.abs(nums[l] + nums[k] + nums[r] - target);
            if (minDiff > diff){
                minDiff = diff;
                ans = nums[l] + nums[k] + nums[r];
            }
        }
        return ans;
    }
    public static void main(String[] args) {
        //int[] nums = {4,0,5,-5,3,3,0,-4,-5};
        int[] nums2 = {1,1,1,1};
        System.out.println("Ans = " + threeSumClosest(nums2, 3));
    }
}
/*
建議先看過第15題
015-three-sum/Solution.java

思路:
只要我們在每次 l ,k, r 要移動時都做下面這個判定, 就能確保最接近 target 的合, 被 minDiff 追蹤著
diff = Math.abs(nums[l] + nums[k] + nums[r] - target);
if (minDiff > diff){
    minDiff = diff;
    ans = nums[l] + nums[k] + nums[r];
}

k 跟 r 是不允許重合在一起的, 否則 minDiff 會記錄到錯誤的值
故每個迴圈都要判斷 k + 1 !=  r
*/